Tests/BuildTests/MakefileFixtures/testBasics/Makefile.txt

```diff
@@ -0,0 +1,243 @@
+# Follow POSIX standards.
+.POSIX:
+
+# Disable all inference rules.
+.SUFFIXES:
+
+# Compute platform specific variables.
+ifeq ($(shell uname -s), Darwin)
+  CC := $(shell xcrun --sdk macosx --find clang)
+  SDKROOT := $(shell xcrun --sdk macosx --show-sdk-path)
+  SWIFT_EXEC := $(shell xcrun --sdk macosx --find swiftc)
+  FRAMEWORKS_DIR := $(shell xcrun --sdk macosx --show-sdk-platform-path)/Developer/Library/Frameworks
+  EXTRA_SWIFT_FLAGS := -F $(FRAMEWORKS_DIR)
+  EXTRA_CC_FLAGS := -F $(FRAMEWORKS_DIR)
+  TARGET := x86_64-apple-macosx10.10
+  TEST_LINKER_ARGS := -Xlinker -bundle
+  NUM_THREADS := $(shell sysctl -n hw.ncpu)
+  CLANG_PLATFORM_FLAGS := -fobjc-arc -arch x86_64 -mmacosx-version-min=10.10
+else ifeq ($(shell uname -s), Linux)
+  SDKROOT := /
+  CC := $(shell which clang)
+  SWIFT_EXEC := $(shell which swiftc)
+  EXTRA_SWIFT_FLAGS :=
+  EXTRA_CC_FLAGS :=
+  TARGET := x86_64-unknown-linux
+  TEST_LINKER_ARGS := -emit-executable -Xlinker -rpath=$ORIGIN
+  NUM_THREADS := $(shell nproc --all)
+  CLANG_PLATFORM_FLAGS :=
+endif
+
+MAKE_FILE_PATH := $(abspath $(dir $(lastword $(MAKEFILE_LIST))))
+SRCROOT := $(abspath $(MAKE_FILE_PATH)/..)
+BUILD_DIR_PATH := $(SRCROOT)/.build
+CONFIG := debug
+UTILS := $(MAKE_FILE_PATH)/utils.py
+BUILD_DIR := $(BUILD_DIR_PATH)/$(TARGET)/$(CONFIG)
+
+ifeq ($(CONFIG), debug)
+  SWIFT_CONFIG_FLAGS := -Onone -g -enable-testing -DDEBUG
+  CLANG_CONFIG_FLAGS := -g -DDEBUG=1 -O0
+else
+  SWIFT_CONFIG_FLAGS := -O -enable-testing
+  CLANG_CONFIG_FLAGS := -O2
+endif
+
+.PHONY: all
+all: ccore.exe core.exe coretests.exe exec.exe exec.exe pkgpackagetests.exe
+	@echo --done--
+
+.PHONY: clean
+clean: 
+	rm -rf $(BUILD_DIR)
+	@echo --cleaned--
+
+# --------------------------------------------------------------- #
+# Target: CCore
+CCORE_TARGET_NAME := CCore
+CCORE_SRCROOT := $(SRCROOT)/Sources/CCore
+CCORE_INCLUDE_DIR := $(CCORE_SRCROOT)/include
+CCORE_SOURCES := $(wildcard $(CCORE_SRCROOT)/*.c)
+CCORE_TEMP_DIR := $(BUILD_DIR)/$(CCORE_TARGET_NAME).build
+CCORE_OBJS := $(subst $(CCORE_SRCROOT),$(CCORE_TEMP_DIR),$(CCORE_SOURCES:%.c=%.c.o))
+CCORE_INCLUDE_PATHS = 
+
+define CCORE_TARGET_RULE
+SRC = $(1)
+OBJ = $$(subst $(2),$(3),$$(SRC:%.c=%.c.o))
+DEPS = $$(OBJ:.o=.d)
+$$(OBJ):
+	@echo Compile $$@
+	@mkdir -p $$(CCORE_TEMP_DIR)
+	@$$(CC) --sysroot $$(SDKROOT) $$(CLANG_PLATFORM_FLAGS) $$(EXTRA_CC_FLAGS) $$(CLANG_CONFIG_FLAGS) -DSWIFT_PACKAGE=1 -fblocks -fmodules -fmodule-name=$$(CCORE_TARGET_NAME) -I $$(CCORE_INCLUDE_DIR) -fmodules-cache-path=$$(BUILD_DIR)/ModuleCache -MD -MT dependencies -MF $$(DEPS) -c $(1) -o $$@
+endef
+
+$(foreach source,$(CCORE_SOURCES),$(eval $(call CCORE_TARGET_RULE, $(source), $(CCORE_SRCROOT), $(CCORE_TEMP_DIR))))
```

Comment:
This is ok, but a little non-traditional. Normally a rule like this would be expressed with a Make rule... I'm not sure if there is anything here that actually requires it use a Make `define` expansion.

I don't really have a strong preference...

Reply:
Yeah this felt odd to me as well but I couldn't figure out a way to generate rules for each source file without spelling them by hand.